### PR TITLE
[PROPOSAL] Switch to pytest style test classes, use plain asserts

### PR DESCRIPTION
**Pull request by epwalsh**
As far as I can tell, since we're using `pytest` to run tests, there's no benefit of using `unittest.TestCase` as the base class for `AllenNlpTestCase` as opposed to plain [`pytest` style classes](https://docs.pytest.org/en/latest/xunit_setup.html?highlight=setup_class#class-level-setup-teardown).

There is, however, several disadvantages:
1.  [`@pytest.mark.parametrize` does not work with methods on a `unittest.TestCase` class](https://stackoverflow.com/questions/18182251/does-pytest-parametrized-test-work-with-unittest-class-based-tests). Given that parametrizing tests is such a powerful tool, I think this is a major issue.
2. `unittest.TestCase` assertion methods (`self.assertEqual`, `self.assertSetEqual`, etc) are less concise than plain `assert` statements, which are recommended by `pytest`.
3. The `setUp` and `tearDown` methods do not have [PEP-8](https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables) compatible names, and it's not clear when these methods are run based on their names. On the other hand, the `pytest` equivalents (`setup_method` and `teardown_method` or `setup_class` and `teardown_class`) have snake-case names that are perfectly clear as to when they run.

---

I know this looks like a huge PR but it's mostly just replacing `unittest` assertion methods with the `pytest` equivalents.

### CODE REVIEW

**Comment by epwalsh** [comment]
The "Check Models" workflow will fail until the `unittest` assertion methods in the models repo are replaced with the pytest equivalents.

File: allennlp/common/testing/test_case.py [hunk 1]
```diff
@@ -54,37 +54,14 @@ def _cleanup_archive_dir_without_logging(path: str):
...
-
-        return wrapper
-
-    return decorator
-
-
 _available_devices = ["cpu"] + (["cuda"] if torch.cuda.is_available() else [])
-multi_device = parametrize(("device",), [(device,) for device in _available_devices])
+
+multi_device = lambda f: pytest.mark.parametrize("device", _available_devices)(pytest.mark.gpu(f))
```
Comment (epwalsh) [inline allennlp/common/testing/test_case.py:64]: The `pytest.mark.gpu(f)` ensures these tests are ran in the "GPU Checks" workflow

**Comment by dirkgr** [comment]
I have been wanting to do this for a while. It's especially confusing in the models repo, because `AllenNlpTestCase` supplies some functions that don't work there.

File: allennlp/tests/commands/docstring_help_test.py [hunk 2]
```diff
@@ -42,20 +42,16 @@ def test_docstring_help(self):
                 subcommand = match.group(2)
                 actual_output = _subcommand_help_output(subcommand)
 
-                self.assertEqual(
-                    expected_output,
-                    actual_output,
+                assert expected_output == actual_output, (
```
Comment (dirkgr) [inline allennlp/tests/commands/docstring_help_test.py:45]: Will this give me those pretty comparisons, that show immediately where the difference is, even in large data?

Comment (epwalsh) [reply allennlp/tests/commands/docstring_help_test.py:45]: Yes sir!

**Comment by epwalsh** [comment]
> This is a bit different than what I thought you were doing. I thought you were changing everything to top-level functions named `test_*`, with common setup and tear-down in [pytest fixtures](https://docs.pytest.org/en/latest/fixture.html#fixture). The base class `AllenNlpTestCase` would then go away. Any reason not to do that?

I think having a base class is beneficial because it wraps up all of its fixtures. Without it, you would have to import the fixtures (the temp directory and whatever else the base class comes with) to their test module or to a `conftest.py` at the same directory level as their test module, because there is no way of declaring global fixtures that work for all test modules without explicitly importing them.

That said, if you don't need those fixtures then it's just unnecessary overhead. So we could probably just use `def test_*` functions in many places.

File: allennlp/tests/data/fields/sequence_label_field_test.py [hunk 3]
```diff
@@ -65,25 +66,25 @@ def test_sequence_label_field_raises_on_incorrect_type(self):
         with pytest.raises(ConfigurationError):
             _ = SequenceLabelField([[], [], [], [], []], self.text)
 
-    def test_class_variables_for_namespace_warnings_work_correctly(self):
+    def test_class_variables_for_namespace_warnings_work_correctly(self, caplog):
```
Comment (matt-gardner) [inline allennlp/tests/data/fields/sequence_label_field_test.py:69]: Nit: `caplog` is not a great variable name.

File: allennlp/tests/training/learning_rate_schedulers/learning_rate_scheduler_test.py [hunk 4]
```diff
@@ -8,19 +9,19 @@
...
 class LearningRateSchedulersTest(AllenNlpTestCase):
-    def setUp(self):
-        super().setUp()
+    def setup_method(self):
+        super().setup_method()
         self.model = torch.nn.Sequential(torch.nn.Linear(10, 10))
 
     def test_reduce_on_plateau_error_throw_when_no_metrics_exist(self):
-        with self.assertRaises(ConfigurationError) as context:
+        with pytest.raises(ConfigurationError) as context:
```
Comment (matt-gardner) [inline allennlp/tests/training/learning_rate_schedulers/learning_rate_scheduler_test.py:17]: I think this should just be something like `match="learning rate ..."`, instead of `as context`, right?

File: allennlp/tests/data/fields/sequence_label_field_test.py (hunk 3 above)
Comment (epwalsh) [reply allennlp/tests/data/fields/sequence_label_field_test.py:69]: This is actually an out-of-the-box pytest fixture so I unfortunately I think we're stuck with that name

Comment (epwalsh) [reply allennlp/tests/data/fields/sequence_label_field_test.py:69]: https://docs.pytest.org/en/latest/logging.html#caplog-fixture

File: allennlp/tests/training/learning_rate_schedulers/learning_rate_scheduler_test.py (hunk 4 above)
Comment (epwalsh) [reply allennlp/tests/training/learning_rate_schedulers/learning_rate_scheduler_test.py]: Good point. Just updated

File: allennlp/tests/data/fields/sequence_label_field_test.py (hunk 3 above)
Comment (matt-gardner) [reply allennlp/tests/data/fields/sequence_label_field_test.py:69]: Oh, hmm, didn't realize that pytest did this magic. Ok 🤷‍♂️
